pgrx/src/pg_catalog.rs

```diff
@@ -0,0 +1,1212 @@
+                #[inline]
+                #[allow(dead_code)]
+                fn get_attr<T: pgrx::datum::FromDatum>(&self, attribute: u32) -> Option<T> {
+                    unsafe {
+                        let mut is_null = true;
+                        let datum = pgrx::pg_sys::SysCacheGetAttr(
+                            self.cache_id,
+                            std::ptr::addr_of!(*self.inner).cast_mut(),
+                            attribute as _,
+                            &mut is_null,
+                        );
+                        T::from_datum(datum, is_null)
+                    }
+                }
+            }
+
+            #[doc = concat!("The search result for pg_catalog.", stringify!($table), ".")]
+            pub struct [<$table:camel Search>] {
+                inner: Option<std::ptr::NonNull<pgrx::pg_sys::HeapTupleData>>,
+                cache_id: i32,
+            }
+
+            impl [<$table:camel Search>] {
+                /// Returns true if the search result is empty.
+                pub fn is_empty(&self) -> bool {
+                    self.inner.is_none()
+                }
+                /// Get the search result.
+                pub fn get(&self) -> Option<[<$table:camel>]> {
+                    unsafe {
+                        Some([<$table:camel>] {
+                            inner: self.inner?.as_ref(),
+                            cache_id: self.cache_id,
+                        })
+                    }
+                }
+            }
+
+            impl Drop for [<$table:camel Search>] {
+                fn drop(&mut self) {
+                    unsafe {
+                        if let Some(inner) = self.inner {
+                            pgrx::pg_sys::ReleaseSysCache(inner.as_ptr());
+                        }
+                    }
+                }
+            }
+
+            #[doc = concat!("The search results for pg_catalog.", stringify!($table), ".")]
+            pub struct [<$table:camel SearchList>] {
+                inner: std::ptr::NonNull<pgrx::pg_sys::CatCList>,
+                cache_id: i32,
+            }
+
+            impl [<$table:camel SearchList>] {
+                /// Returns the number of the search results.
+                pub fn len(&self) -> usize {
+                    unsafe {
+                        let inner = self.inner.as_ref();
+                        inner.n_members as usize
+                    }
+                }
+                /// Returns true if the search result is empty.
+                pub fn is_empty(&self) -> bool {
+                    self.len() == 0
+                }
+                /// Get the `i`-th search result.
+                pub fn get(&self, i: usize) -> Option<[<$table:camel>]> {
+                    unsafe {
+                        let inner = self.inner.as_ref();
+                        let slice = inner.members.as_slice(inner.n_members as usize);
+                        let member = *slice.get(i)?;
```

Comment:
This is an example of where the types get sufficiently confusing without any annotations. I understand if there are cases where the types are, due to the macro expansion, sufficiently generic that it is basically implausible to type annotate them at that point, however. Only add them where it is easy and clarifying to do so.